File: Add Code Here/C++/tree/Sorting Algorithms Using C/mergesort.c
#include<stdio.h>
#include<conio.h>
void Mergesort(int[],int,int);
void Merge(int[],int ,int,int);
int main()
{
    int arr[50];
    int i,size;
    printf("Enter the size of array");
    scanf("%d",&size);
    printf("Enter the elements");
    for(i=0;i<size;i++)
        scanf("%d",&arr[i]);
    Mergesort(arr,0,size-1);
    printf("After Merge Sort");
    for(i=0;i<size;i++)
        printf("%d ",arr[i]);
    return 0;
}
void Mergesort(int arr[],int low,int high)
{
    int mid;
    if(low<high)
    {
        mid=(low+high)/2;
        Mergesort(arr,low,mid);
        Mergesort(arr,mid+1,high);
        Merge(arr,low,mid,high);
    }
}
void Merge(int arr[],int low,int mid,int high)
{
    int i,j,k,lo,temp[50];
    lo=low;
    i=low;
    j=mid+1;
    while(i<=mid && j<=high)
    {
        if(arr[i]<=arr[j])
        {
            
            temp[lo]=arr[i];
            i++;
        }
        else
        {
            temp[lo]=arr[j];
            j++;
        }
        lo++;
        
    }
    if(i>mid)
    {
        for(k=j;k<=high;k++)
        {
            temp[lo]=arr[k];
            lo++;

        }
    }
    else
    {
        for(k=i;k<=mid;k++)
        {
         temp[lo]=arr[k];
         lo++;  
        }
    }
    for(k=0;k<=high;k++)
        arr[k]=temp[k];
    
}

File: Add Code Here/JAVA/Sorting/MergeSort.java
// Java program to sort an array
// using Merge sort
import java.util.*;
public class MergeSort {
    public static void main(String[] args) {
        
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter the size of array:");
        int n=sc.nextInt();
        int arr[]=new int[n];
        System.out.println("Enter the element:");
        for(int i=0;i<n; i++){
            arr[i]=sc.nextInt();
        }

        divide(arr,0, n-1);
        System.out.println("Sorted Array:");
        for(int i=0; i<n; i++){
            System.out.println(arr[i]+" ");
        }
        System.out.println();
    }

    public static void divide(int arr[], int si, int ei){

        if(si>=ei){
            return;
        }
        int mid=si + (ei - si)/2;
        divide(arr, si, mid);
        divide(arr, mid+1, ei);
        conquer(arr, si, mid, ei);
        
    }
    public static void conquer(int arr[], int si, int mid, int ei){
        int merged[] = new  int [ei-si+1];           //[1,4,2,2]  indx=  3 - 0 +1 =4 (size of arr) 

        int idx1=si;
        int idx2=mid+1;
        int x=0;

        while(idx1 <=mid && idx2 <=ei){
            if(arr[idx1] <= arr[idx2]){
                merged[x] = arr[idx1];
                x++;
                idx1++;
            }else{
                merged[x]=arr[idx2];
                x++;
                idx2++;
            }
        }
        while(idx1<=mid){
            merged[x]=arr[idx1];
            x++;
            idx1++;
        }
        while(idx2<=ei){
            merged[x]=arr[idx2];
            x++;
            idx2++;
        }
        for(int i=0, j=si; i<merged.length; i++, j++){
            arr[j]=merged[i];
        }
        
    }
}

// This code is contributed by Md Tajuddin
/*
OUTPUT: 

Enter the size of array:5
Enter the element:
3 5 2 7 1
Sorted Array:
1 
2
3
5
7
 */